From 3f76418eaf73896489129c529fac021e4f3a03c0 Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@gentoo.org>
Date: Wed, 21 Nov 2018 11:00:54 +0100
Subject: [PATCH] use awk/sed instead of perl for creating header files

More systems should have awk/sed than perl.

Signed-off-by: Mike Frysinger <vapier@gentoo.org>

Forward ported from libcap-2.22 to libcap-2.26
and incorporated the gperf-3.1 fix provided by Mike Gilbert
<floppym@gentoo.org>
Forward ported from libcap-2.26 to libcap-2.28
Forward ported from libcap-2.28 to libcap-2.38

Signed-off-by: Lars Wendler <polynomial-c@gentoo.org>
---
 libcap/Makefile | 26 +++++++++++++++++++-------
 1 file changed, 19 insertions(+), 7 deletions(-)

diff --git a/libcap/Makefile b/libcap/Makefile
index 3d6463d..df2526c 100644
--- a/libcap/Makefile
+++ b/libcap/Makefile
@@ -22,6 +22,8 @@ MAJLIBNAME=$(LIBNAME).$(VERSION)
 MINLIBNAME=$(MAJLIBNAME).$(MINOR)
 GPERF_OUTPUT = _caps_output.gperf
 CFLAGS += -fPIC
+AWK = awk
+SED = sed
 
 all: $(MINLIBNAME) $(STACAPLIBNAME) libcap.pc libpsx.pc $(STAPSXLIBNAME)
 
@@ -55,17 +57,27 @@ cap_names.h: _makenames
 	./_makenames > cap_names.h
 
 $(GPERF_OUTPUT): cap_names.list.h
-	perl -e 'print "struct __cap_token_s { const char *name; int index; };\n%{\nconst struct __cap_token_s *__cap_lookup_name(const char *, size_t);\n%}\n%%\n"; while ($$l = <>) { $$l =~ s/[\{\"]//g; $$l =~ s/\}.*// ; print $$l; }' < $< | gperf --ignore-case --language=ANSI-C --readonly --null-strings --global-table --hash-function-name=__cap_hash_name --lookup-function-name="__cap_lookup_name" -c -t -m20 $(INDENT) > $@
-	sed -e 's/unsigned int len/size_t len/' -i $@
+	(printf "%b" "struct __cap_token_s { const char *name; int index; };\n%%\n"; \
+	$(SED) -e 's:["{}]::g' -e 's:,$$::' $<) | \
+	gperf \
+		--ignore-case \
+		--language=ANSI-C \
+		--includes \
+		--readonly \
+		--null-strings \
+		--global-table \
+		--hash-function-name=__cap_hash_name \
+		--lookup-function-name="__cap_lookup_name" \
+		-c -t -m20 $(INDENT) > $@
 
-# Intention is that libcap keeps up with torvalds' tree, as reflected
-# by this maintained version of the kernel header. libcap dynamically
-# trims the meaning of "all" capabilities down to that of the running
-# kernel as of 2.30.
 UAPI_HEADER := $(topdir)/libcap/include/uapi/linux/capability.h
 cap_names.list.h: Makefile $(UAPI_HEADER)
 	@echo "=> making $@ from $(UAPI_HEADER)"
-	perl -e 'while ($$l=<>) { if ($$l =~ /^\#define[ \t](CAP[_A-Z]+)[ \t]+([0-9]+)\s+$$/) { $$tok=$$1; $$val=$$2; $$tok =~ tr/A-Z/a-z/; print "{\"$$tok\",$$val},\n"; } }' $(UAPI_HEADER) | fgrep -v 0x > $@
+	$(AWK) '($$0 ~ /^#define[[:space:]]+CAP[_A-Z]+[[:space:]]+[0-9]+[[:space:]]*$$/) { printf "{\"%s\",%s},\n", tolower($$2), $$3 }' $(UAPI_HEADER) > $@
+
+cap_names.list.h: $(KERNEL_HEADERS)/linux/capability.h Makefile
+	@echo "=> making $@ from $<"
+	$(AWK) '($$0 ~ /^#define[[:space:]]+CAP[_A-Z]+[[:space:]]+[0-9]+[[:space:]]*$$/) { printf "{\"%s\",%s},\n", tolower($$2), $$3 }' $< > $@
 
 $(STACAPLIBNAME): $(CAPOBJS)
 	$(AR) rcs $@ $^
-- 
2.27.0

